versions.lock

```diff
@@ -20,7 +20,6 @@ com.google.code.gson:gson:2.2.4 (2 constraints: 9518bfd2)
 com.google.errorprone:error_prone_annotations:2.3.3 (2 constraints: 161a2544)
 com.google.flatbuffers:flatbuffers-java:1.9.0 (2 constraints: e5199714)
 com.google.guava:failureaccess:1.0.1 (1 constraints: 140ae1b4)
-com.google.guava:guava:28.0-jre (21 constraints: 88453dad)
```

Comment:
I couldn't figure out how to have Guava removed entirely from the `versions.lock` file so I removed it manually. I tried the following unsuccessfully:

* Mark guava in `iceberg-relocate-external-dependencies` as `compileOnly` (which to my limited understanding of Gradle is the equivalent to Maven's `provided) - this unfortunately results in the classes not being seen by the shadow plugin so they don't end up in the jar and Guava 28 still ends up in `versions.lock`
* Only apply `com.palantir.consistent-versions` to other sub-projects (this doesn't seem to work as the plugin refuses to be set anywhere other than on the root project)

Is there some other way to tell the versions plugin to ignore the Guava dependency in `iceberg-relocate-external-dependencies`? Ideally we'd "ban" Guava entirely to indicate to developers they need to use the relocated version.

Reply:
I think it is `./gradlew --write-locks`. But I'm not sure that we need to exclude this from our locks. We should only need to exclude it from direct dependencies. As long as it isn't showing up as a transitive dependency of `iceberg-relocate-external-dependencies`, we should be okay.

We will also need to add a LICENSE and NOTICE file for the shaded Jar, the same way that we do for `iceberg-spark-runtime`

Reply:
Two more things:

For modules that need a dependency on a different guava version, does excluding the transitive dependency work?

What do you think about keeping guava dependencies for modules that haven't been moved? Then we could migrate one module at a time, which would cut down on commit conflicts.

Reply:
> We will also need to add a LICENSE and NOTICE file for the shaded Jar, the same way that we do for `iceberg-spark-runtime`

That's happening on lines 120-123 no?

Reply:
> I think it is `./gradlew --write-locks`. But I'm not sure that we need to exclude this from our locks. We should only need to exclude it from direct dependencies. As long as it isn't showing up as a transitive dependency of `iceberg-relocate-external-dependencies`, we should be okay.

That just re-writes the locks, and I think the plugin then sees that this module is using Guava 28 so it puts that in because it considers it a transitive dependency like you say. If in iceberg-hive we put Guava 14 then it still chooses 28 for versions.lock and this overrides the version used in iceberg-hive. My understanding is that this consistent versions thing doesn't let you choose a version per subproject put perhaps there is some magic that can allow that?

Also, this might work if I apply this change to all the subprojects and ensure that none of them is transitively pulling in Guava 28 and then set iceberg-hive to the expected version. So perhaps let's sort out the other questions on this PR about the package name etc. and I can then apply that everywhere and we can take it from there?

Reply:
> Two more things:
> 
> For modules that need a dependency on a different guava version, does excluding the transitive dependency work?

In order to get this working in [our other branch ](https://github.com/ExpediaGroup/incubator-iceberg/blob/shade-dependencies-aw/build.gradle) we had to do both. First we shaded Guava everywhere so that the consistent versions plugin wouldn't choose a higher version. Second we then also had to exclude transitive versions of any other version of Guava. Then finally it would end up with the correct one for the iceberg-hive classpath. This was very simple to do with Maven (we tried in a separate project) and didn't require many of these tricks so maybe we're missing some feature in Gradle or consistent versions that allows you to do what you describe?

> 
> What do you think about keeping guava dependencies for modules that haven't been moved? Then we could migrate one module at a time, which would cut down on commit conflicts.

I think the problem with the above is that the consistent versions plugin will look at any Guava dependency and then choose the highest version and then enforce that across the entire project and subprojects. We could roll this change out one module at a time but it won't be usable in `iceberg-hive` until it's been applied everywhere and we can confirm that Guava 28 really is out of the picture.

Reply:
In the other shaded module, we shade the `compileOnly` configuration and add the dependencies to shade there. That way, the dependencies are not listed in the final artifact. You could try that if you want, but the current setup looks fine to me. I don't see guava in the other modules.

Reply:
Also, I expect to have a guava entry in versions.lock because it is listed as a dependency. As long as it is the expected version, we're okay.